source/Symbol/SwiftASTContext.cpp

```diff
@@ -8032,3 +8032,173 @@ void lldb_private::printASTValidationInfo(
+    lldb::StackFrameSP this_frame_sp(stack_frame_wp.lock());
+
+    if (module_name == ConstString(this->GetClangImporter()
+                                       ->getImportedHeaderModule()
+                                       ->getName()
+                                       .str()))
+      swift_module = this->GetClangImporter()->getImportedHeaderModule();
+    else if (this_frame_sp) {
+      lldb::ProcessSP process_sp(this_frame_sp->CalculateProcess());
+      if (process_sp)
+        swift_module = this->FindAndLoadModule(module_name, *process_sp.get(),
+                                               error);
+    } else
+      swift_module = this->GetModule(module_name, error);
+
+    if (!swift_module || !error.Success() || this->HasFatalErrors()) {
+      if (log)
+        log->Printf("[PerformAutoImport] Couldn't import module %s: %s",
+                    module_name.AsCString(), error.AsCString());
+      return false;
+    }
+
+    if (log) {
+      log->Printf("Importing %s with source files:", module_name.AsCString());
+
+      for (swift::FileUnit *file_unit : swift_module->getFiles()) {
+        StreamString ss;
+        DescribeFileUnit(ss, file_unit);
+        log->Printf("  %s", ss.GetData());
+      }
+    }
+
+    additional_imports.push_back(swift::SourceFile::ImportedModuleDesc(
+        std::make_pair(swift::ModuleDecl::AccessPathTy(), swift_module),
+        swift::SourceFile::ImportOptions()));
+    return true;
+  };
+
+  if (!user_imports) {
+    if (!load_one_module(ConstString("Swift")))
+      return false;
+    if (cu_modules) {
+      for (const ConstString &module_name : *cu_modules) {
+        if (!load_one_module(module_name))
+          return false;
+      }
+    }
+  } else {
+    llvm::SmallVector<swift::ModuleDecl::ImportedModule, 2> parsed_imports;
+    source_file.getImportedModules(parsed_imports,
+                                   swift::ModuleDecl::ImportFilter::All);
+    auto *persistent_expression_state =
+        sc.target_sp->GetSwiftPersistentExpressionState(exe_scope);
+
+    for (auto module_pair : parsed_imports) {
+      swift::ModuleDecl *module = module_pair.second;
+      if (!module)
+        continue;
+      std::string module_name;
+      GetNameFromModule(module, module_name);
+      if (module_name.empty())
+        continue;
+      ConstString module_const_str(module_name);
+      if (log)
+        log->Printf("[PerformAutoImport] Performing auto import on found "
+                    "module: %s.\n",
+                    module_name.c_str());
+      if (!load_one_module(module_const_str))
+        return false;
+      // How do we tell we are in REPL or playground mode?
+      persistent_expression_state->AddHandLoadedModule(module_const_str);
+    }
+    // Finally get the hand-loaded modules from the
+    // SwiftPersistentExpressionState and load them into this context:
+    if (!persistent_expression_state->RunOverHandLoadedModules(load_one_module))
```

Comment:
Looking at the implementation of RunOverHandLoadedModules this is just a really fancy `for`-loop. I wonder where else we need `m_hand_loaded_modules` if at all.

Reply:
We are currently using m_hand_loaded_modules in two ways.  

The first is in PerformAutoImport where we dig through the parsed AST and find all the import statements and then use RunOverHandLoadedModules in the same function to import them.  

We also use it to force loading the Swift module so that you can use the swift expression parser even if you haven't loaded Swift yet (e.g. in the REPL), for instance in ParseAndImport.  That was really what the ivar was for, so various different pieces of code could say "I need you to load this for me!" wherever you were in the expression parser and it would get done.  

OTOY, that only got used to force Swift to be loaded.  Since I don't think you're ever going to get very far in the swift expression parser w/o loading the Swift module, if it's cheap to "import" the Swift module when it's already present, maybe PerformAutoImport could just always do that?

Reply:
The bits with the persistent expression state might be better if they remained in the expression parser, since they won't be needed by anyone else.